### PR TITLE
Density serialization

### PR DESCRIPTION
**Pull request by jmalkin**
Adds density serialization

### CODE REVIEW

File: density/include/density_sketch_impl.hpp [hunk 1]
```diff
@@ -146,6 +158,253 @@ void density_sketch<T, K, A>::compact_level(unsigned height) {
...
+  uint8_t flags_byte;
+  ptr += copy_from_mem(ptr, flags_byte);
+  uint16_t k;
+  ptr += copy_from_mem(ptr, k);
+  uint16_t unused;
+  ptr += copy_from_mem(ptr, unused);
+  uint32_t dim;
+  ptr += copy_from_mem(ptr, dim);
+
+  //check_k(k);
```
Comment (AlexanderSaydakov) [inline density/include/density_sketch_impl.hpp:329]: do we need to check K here?

Comment (jmalkin) [reply density/include/density_sketch_impl.hpp:329]: We don't check anything in the constructor so I dunno what the value range would be? It's uint16_t so it obviously can't be negative.  We can check that it's strictly positive?  It seems like there should be some minimum practical value given the need for compression, though.

Comment (AlexanderSaydakov) [reply density/include/density_sketch_impl.hpp:329]: I am not sure we need to check, but I saw this commented-out line, which brings a question whether to remove it or implement.

Comment (jmalkin) [reply density/include/density_sketch_impl.hpp:329]: My top choice would be to check. I'm not sure of a practical minimum though. Fairly sure 1 is not a good idea, and 0 should not be allowed. So maybe uncomment, check >1, and also add to the constructor?

Comment (AlexanderSaydakov) [reply density/include/density_sketch_impl.hpp:329]: I would agree to that, if there are no other practical limits

File: density/include/density_sketch_impl.hpp [hunk 2]
```diff
@@ -146,6 +162,259 @@ void density_sketch<T, K, A>::compact_level(unsigned height) {
...
+
+  if (is_empty())
+    return;
+
+  write(os, num_retained_);
+  write(os, n_);
+
+  // levels array -- uint32_t since a single level may be larger than k
+  size_t pt_size = sizeof(T) * dim_;
+  for (Level lvl : levels_) {
```
Comment (AlexanderSaydakov) [inline density/include/density_sketch_impl.hpp:208]: shouldn't this be a (const) ref?

File: density/include/density_sketch_impl.hpp [hunk 3]
```diff
@@ -146,6 +162,259 @@ void density_sketch<T, K, A>::compact_level(unsigned height) {
...
+    const uint32_t level_size = static_cast<uint32_t>(lvl.size());
+    write(os, level_size);
+    for (Vector pt : lvl) {
```
Comment (AlexanderSaydakov) [inline density/include/density_sketch_impl.hpp:211]: and here const ref perhaps?

File: density/include/density_sketch_impl.hpp [hunk 4]
```diff
@@ -146,6 +162,259 @@ void density_sketch<T, K, A>::compact_level(unsigned height) {
...
+  
+  if (is_empty())
+    return bytes;
+
+  ptr += copy_to_mem(num_retained_, ptr);
+  ptr += copy_to_mem(n_, ptr);
+
+  // levels array -- uint32_t since a single level may be larger than k
+  size_t pt_size = sizeof(T) * dim_;
+  for (Level lvl : levels_) {
```
Comment (AlexanderSaydakov) [inline density/include/density_sketch_impl.hpp:252]: const ref

File: density/include/density_sketch_impl.hpp [hunk 5]
```diff
@@ -146,6 +162,259 @@ void density_sketch<T, K, A>::compact_level(unsigned height) {
...
+    ptr += copy_to_mem(static_cast<uint32_t>(lvl.size()), ptr);
+    for (Vector pt : lvl) {
```
Comment (AlexanderSaydakov) [inline density/include/density_sketch_impl.hpp:254]: const fer

**Review by AlexanderSaydakov — APPROVED** [review]

**Pull request merged**